content/blog/2020-10-04-alive/index.md

```diff
@@ -0,0 +1,227 @@
+
+```
+%z = or i8 1, undef
+```
+
+where the value of `%z` becomes any odd integer in the range `{0, ..., 255}`. `undef` allows the compiler to aggresively optimize a program as LLVM [makes the assumption](undef-val) that whenever an `undef` is seen "the program is well-defined no matter what value is used."
+
+`poison` values are distinct from `udef` values as they are used to indicate that "a side-effect-free instruction has a condition that produces undefined behavior." There is no way to explicitly indicate that a value is `poison` in LLVM. As an example, the LLVM instruction 
+
+```
+%r = shl nsw %x, log2(C1)
+```
+
+causes `%r` to be a `poison` value as the left shift might cause `%x` to go from a positive value to a negative one. If the value of `%r` were to be used in an instruction with side-effects, such as memory stores, then we get true undefined behavior. Furthermore, `poison` values will taint any subsequent dependent instructions meaning `poison` is propagated throughout a program. `poison` values are deferred undefined behaviors and can only be identified through careful analyses.
+
+Finally, instruction attributes allow the compiler to make assumptions about certain instructions. For example, `nsw` means "no signed wrap" which makes signed overflow undefined allowing us to perform the optimization where `x + 1 > x` is replaced with `true`.
+
+### Refinement
+
+By considering the types of undefined behavior in LLVM, the authors define what valid refinements of programs are. If there is an `undef` value then an optimization must produce a value that is subset of the `undef` value produced by the source as `undef` represents a set of possible values. Similarly, if the source program has a `poison` value then the target program may have a `poison` value but a target instruction cannot create `poison` values when there were none in the source program. Instruction attributes indicate when an instruction can be considered undefined behavior and when it must be considered as not being undefined (such as signed overflow vs unsigned overflow). So, the authors use the intuition that compilers should not introduce new undefined behavior to generate SMT formulas that capture the following correctness criteria:
+
+> (1) the target is defined when the source is defined,
+>
+> (2) the target is poison-free when the source is poison-free, and
+>
+> (3) the source and the target produce the same result when the source is defined and poison-free.
+
+The SMT formulas generated by Alive encode all possible types for the instructions in the source and target programs and also encode `undef`, `poison` and instruction attributes to model undefined behavior. Three main formulas corresponding to the correctness criterias above are generated and passed to the SMT solver for verification.
+
+### Example
+
+The optimization of replacing `(X << C1) / C2 ` with `X / (C2 >> C1)` whenever `C2 is a multiple of C1` can be written in Alive as follows
```

Comment:
No need for backticks around that English condition.

Reply:
Fixed.